web_poet/pages.py

```diff
@@ -68,6 +68,25 @@ async def to_item(self) -> ItemT:
         )
 
 
+class MultiLayoutPage(ItemPage[ItemT]):
```

Comment:
I also wonder if we should support fields for MultiLayoutPage, or not. 
Currently all fields defined for MultiLayoutPage are silently ignored. 

There is also a use case when you need a "partial" layout, layout of a region of a page. In this case, only some fields are extracted by the layout; others are common. It seems we have a few options:

1. Use a base class with common fields. Create layouts which inherit from the base class and handle necessary regions. Use them in MultiLayoutPage.
2. Allow to define fields on the MultiLayoutPage. When layout is picked, combine the fields from the layout with the fields defined in the class itself, to compute the final item. Supporting to_item for layouts can be more tricky in this case, as it's not clear which item they should use.

A separate issue is if it's possible to use 2 or more regions, with different layouts, in the same page object.

If fields are supported, it seems it makes sense to move the logic to ItemPage. I think it may simplify typing, and inheritance as well. E.g. layouts can be used with ProductPage from https://github.com/zytedata/zyte-common-items/pull/19.

It seems that if fields are not supported, it's better to keep MultiLayoutPage as a separate class, and probably raise an error or issue a warning if fields are defined.

Sorry for a braindump :) What do you think?

Reply:
A disadvantage of using MultiLayoutPage without fields: it's not possible to use fields in the code which uses the page object. So let's say we have ProductPage, it uses fields for data extraction. Then, it's refactored to use MultiLayoutPage as a base class. It means that the fields are no longer supported, and so the code which uses this page may break.

Reply:
I am slightly against field support in MultiLayoutPage.

In addition to those 2 options you mention, I can think of a 3rd that is a variation on 1 based precisely on how you amended my API proposal for MultiLayoutPage:

3. Use composition rather than inheritance: define a page object class that can extract the common fields, make it a dependency of the layout(s) that extract additional fields, and use it in that layout.

We would still have the same problem as with the lack of fields in MultiLayoutPage, i.e. you could not access the fields of the dependency layout through the layout that uses it (other than accessing the dependency directly, e.g. `layout.dependency.field`). But I wonder if we could implement a getattr fallback mechanism to solve this issue for both scenarios: allow MultiLayoutPage to expose the fields of the object that its layout() method returns, and allow layouts with other layouts as dependencies to expose the fields of their dependencies.

Reply:
I'm also not sure about support `@fields` in `MultiLayoutPage`.

It would seem it'd be best to keep it's task simple wherein it simply identifies and returns the PO instance based on the layout.

Reply:
> There is one argument for keeping it separate and not supporting fields: it'd allow to define fields named `layout`. If we provide a standard method named `layout`, and allow to define fields, it's not possible to have a field of the same name.

This is a valid point, but I think we could switch to ``get_layout`` to avoid this issue, and it would be a better method name anyway.

Reply:
I still think defining fields on a multi-layout page object should be discouraged, but the more I think about it, the more I feel we indeed need to allow addressing the suggested refactoring scenario:

> So let's say we have ProductPage, it uses fields for data extraction. Then, it's refactored to use MultiLayoutPage as a base class. It means that the fields are no longer supported, and so the code which uses this page may break.

With the current proposal, you would have to manually define every field as:

```python
        @field
        async def foo(self) -> Optional[str]:
            return await self.layout().foo
```

When I wrote [my earlier comment](https://github.com/scrapinghub/web-poet/pull/103/files#r1018112627), I was thinking that we could make `__getattr__` fall back to `self.layout().__getattr__` to address this, so we do not need to add code like this to every field.

But when you inherit from `MultiLayoutPage`, other issues arise. For example, the `to_item` method of the inheriting class becomes that of `MultiLayoutPage`, preventing fields from the inheriting class to be included in the output of `to_item`.

Reply:
I wonder if, rather than providing a `MultiLayoutPage`, we should document how to implement one instead by:
- Defining a cached, class-private ``__get_layout`` method for the layout selection code, so that subclasses can safely redefine it, for cases of multiple multi-layout classes in the inheritance tree.
- Defining every single field from the underlying layout as in my previous comment. It is tedious, but it also plays well with `ItemPage.to_item`.

Reply:
@Gallaecio can you please explain the last idea in more details, e.g. with some code?

I agree that not having fields in `MultiLayoutPage` is a better design and I also agree that there are problems with it.

Reply:
> @Gallaecio can you please explain the last idea in more details, e.g. with some code?

https://github.com/scrapinghub/web-poet/pull/103/commits/fc7867fb0d9a8c48cc43d3bf2ec295d09fbf724f

Reply:
Yeah, this looks nice.

Reply:
We were discussing it today on a call with @proway2. They implemented a multi-layout page object, tried a few approaches. In short - having fields on the final page object is a must :) That's the reason the documented approach here won't work well for them. Taking union of all dependencies is fine.

Reply:
Currently they define all the fields, and call to the self._layout in each field. It's a lot of boilerplate; exactly something a library should be solving.

Reply:
It seems having fields availabe is required, but not necessarily being able to define top-level fields.